### PR TITLE
[TorchToTosa] Fix i32 and i64 aten.clamp lowering

### PR DESCRIPTION
**Pull request by jackv1ck**
Tosa clamp does not support i32 and i64 operands so lowering aten.clamp directly to tosa.clamp results in invalid tosa for these values.

This PR updates the lowering for scalar aten.clamp `AtenClampOp` so i8 and i16 operands continue to use tosa.clamp, while i32 and i64 are lowered as `y = min(max(x, min_value), max_value)`, in the same way as the existing lowering of the tensor aten.clamp `AtenClampTensorOp`

### CODE REVIEW

File: lib/Conversion/TorchToTosa/TorchToTosa.cpp [hunk 1]
```diff
@@ -5991,13 +5991,44 @@ LogicalResult ConvertAtenOp<AtenIscloseOp>::matchAndRewriteImpl(
...
+  auto minThresholdCheck = tosa::MaximumOp::create(
+      rewriter, op->getLoc(), resultType, self, min,
+      /*nan_mode=*/
+      tosa::NanPropagationModeAttr::get(rewriter.getContext(),
+                                        tosa::NanPropagationMode::PROPAGATE));
+
+  // yi = min(max(xi, min_valuei), max_valuei)
+  // Use default NaN Propagation mode "PROPAGATE" for tosa.minimum
+  auto result = tosa::MinimumOp::create(
+      rewriter, op->getLoc(), resultType, minThresholdCheck, max,
```
Comment (Lallapallooza) [inline lib/Conversion/TorchToTosa/TorchToTosa.cpp]: Can we avoid using the final clamp result type for the intermediate `tosa.maximum` here?

For `aten.clamp.Tensor`, a valid broadcast like `self: [3]`, `min: []`, `max: [2,3]`, result `[2,3]` makes `maximum(self, min)` broadcast only to `[1,3]`, so creating it with result type `[2,3]` is invalid

File: lib/Conversion/TorchToTosa/TorchToTosa.cpp [hunk 2]
```diff
@@ -6055,34 +6085,70 @@ LogicalResult ConvertAtenOp<AtenClampOp>::matchAndRewriteImpl(
...
+        //    yi = min(max(xi, min_valuei), max_valuei)
+        switch (intType.getWidth()) {
+        case 8:
+        case 16:
+          rewriter.replaceOpWithNewOp<tosa::ClampOp>(
+              op, outType, self, minIntAttr, maxIntAttr,
+              /*nan_mode=*/
+              tosa::NanPropagationModeAttr::get(
+                  rewriter.getContext(), tosa::NanPropagationMode::PROPAGATE));
+          return success();
```
Comment (Lallapallooza) [inline lib/Conversion/TorchToTosa/TorchToTosa.cpp:6163]: This keeps i8/i16 scalar clamp on `tosa.clamp`, which rejects `min_val > max_val`, while PyTorch accepts `torch.clamp(x, min=4, max=2)` and returns all `2`s.

Since the new i32/i64 max/min decomposition already matches PyTorch for that case, can we either use that decomposition for all integer widths or at least route small-width `min > max` through it?

File: lib/Conversion/TorchToTosa/TorchToTosa.cpp [hunk 3]
```diff
@@ -6055,34 +6085,70 @@ LogicalResult ConvertAtenOp<AtenClampOp>::matchAndRewriteImpl(
...
-      return failure();
-    }
-
-    rewriter.replaceOpWithNewOp<tosa::ClampOp>(
-        op, outType, self, minFloatAttr, maxFloatAttr,
-        /*nan_mode=*/
-        tosa::NanPropagationModeAttr::get(rewriter.getContext(),
-                                          tosa::NanPropagationMode::PROPAGATE));
-  }
+  return TypeSwitch<Type, LogicalResult>(outElemTy)
```
Comment (Lallapallooza) [inline lib/Conversion/TorchToTosa/TorchToTosa.cpp:6137]: Can we validate explicit scalar integer bounds against the destination integer type before creating attrs/constants and fail the match if they are not representable?

File: lib/Conversion/TorchToTosa/TorchToTosa.cpp
Comment (Lallapallooza) [inline lib/Conversion/TorchToTosa/TorchToTosa.cpp:6246]: Can we add case 16 here?

Comment (Lallapallooza) [inline lib/Conversion/TorchToTosa/TorchToTosa.cpp:6285]: Same, can we add case 16 here?

File: test/Conversion/TorchToTosa/basic.mlir [hunk 4]
```diff
@@ -1546,15 +1546,87 @@ func.func @torch.aten.slice.bool_strided(%arg0: !torch.vtensor<[1,64,1],i1>) ->
   return %0 : !torch.vtensor<[1,32,1],i1>
 }
 
+// -----
```
Comment (Lallapallooza) [inline test/Conversion/TorchToTosa/basic.mlir:1549]: Can we add coverage for the edge cases exercised by this helper, not just the same-shape scalar happy path? In particular, an `si16` scalar case would lock down the `case 16` side of the width split, and an integer `aten.clamp.Tensor` broadcast case such as `self [3]`, `min []`, `max [2,3]` would catch the invalid intermediate `tosa.maximum` result type.

**Review by Lallapallooza — CHANGES_REQUESTED** [review]
Thanks for the patch. Few comments, could you also please add an e2e test as well.

File: lib/Conversion/TorchToTosa/TorchToTosa.cpp (hunk 3 above)
Comment (jackv1ck) [reply lib/Conversion/TorchToTosa/TorchToTosa.cpp:6137]: Done

File: lib/Conversion/TorchToTosa/TorchToTosa.cpp (hunk 2 above)
Comment (jackv1ck) [reply lib/Conversion/TorchToTosa/TorchToTosa.cpp:6163]: Done. Set min to max when min is larger

File: test/Conversion/TorchToTosa/basic.mlir (hunk 4 above)
Comment (jackv1ck) [reply test/Conversion/TorchToTosa/basic.mlir:1549]: Done

File: lib/Conversion/TorchToTosa/TorchToTosa.cpp
Comment (jackv1ck) [reply lib/Conversion/TorchToTosa/TorchToTosa.cpp:6246]: Done

Comment (jackv1ck) [reply lib/Conversion/TorchToTosa/TorchToTosa.cpp:6285]: Done

File: lib/Conversion/TorchToTosa/TorchToTosa.cpp (hunk 1 above)
Comment (jackv1ck) [reply lib/Conversion/TorchToTosa/TorchToTosa.cpp]: Done

**Review by Lallapallooza — APPROVED** [review]
LGTM
